execution_workflows/APAtrap/bin/convert_to_bed.py

```diff
@@ -0,0 +1,62 @@
+#!/usr/bin/env python3
+
+import sys
+import argparse
+import pandas as pd
+
+def parse_args(args=None):
+    Description = "Reformat APAtrap identifyDistal3UTR bed file into the output files of identification and quantification challenges"
+    Epilog = "Example usage: python convert_to_tsv.py <FILE_IN> <IDENTIFICATION_OUT> <QUANTIFICATION_OUT>"
+
+    parser = argparse.ArgumentParser(description=Description, epilog=Epilog)
+    parser.add_argument("FILE_IN", help="Input deAPA output txt file.")
+    parser.add_argument("IDENTIFICATION_OUT", help="Name of output file for identification challenge")
+    parser.add_argument("QUANTIFICATION_OUT", help="Name of output file for quantification challenge")
+    return parser.parse_args(args)
+
+
+def reformat_bed(file_in, identification_out, quantification_out):
+    """
+    This function reformats the txt predictAPA output file to files for
+    identification and quantification challenges
+    :param file_in: txt file to be reformatted
+    :return: N/A
+    """
+    identification_out = open(identification_out, "wt")
+    quantification_out = open(quantification_out, "wt")
+
+    df = pd.read_csv(file_in, sep='\t')
+    for index, row in df.iterrows():
+        # replace score column with "."
+        # keep just the gene name
+        name = row['Gene'].split("|")[0]
+        chrom = row['Gene'].split("|")[2]
+        strand = row['Gene'].split("|")[3]
+        apas = row['Predicted_APA'].split(",")
+
+        # write identification file
+        for apa in apas:
+            chromStart = str(apa)
+            # as identified PAS are single-nucleotide, the ending position is the
+            # same as starting position
+            chromEnd = chromStart
+            score = "."
+            output = [chrom, chromStart, chromEnd, name, score, strand]
+            identification_out.write("\t".join(output) + "\n")
+
+            # write quantification file
+            score = str(row['Group_1_1_Total_Exp'])
+            output = [chrom, chromStart, chromEnd, name, score, strand]
+            quantification_out.write("\t".join(output) + "\n")
```

Comment:
my quantification output looks like this (`control_r1`):
| 16 | 78340249  | 78340249  | ENSMUST00000023572.14 | 85.72  | + |
|----|-----------|-----------|-----------------------|--------|---|
| 16 | 78340414  | 78340414  | ENSMUST00000023572.14 | 85.72  | + |
| 6  | 119924719 | 119924719 | ENSMUST00000088646.11 | 161.84 | - |
| 6  | 119924739 | 119924739 | ENSMUST00000159903.1  | 161.91 | - |
| 6  | 119924260 | 119924260 | ENSMUST00000159903.1  | 161.91 | - |
| 6  | 119924739 | 119924739 | ENSMUST00000177761.7  | 161.91 | - |
| 6  | 119924260 | 119924260 | ENSMUST00000177761.7  | 161.91 | - |

It looks like proximal and distal poly(A) sites of the same txipt are given the same quantification value, I think because you're using the total expression column `Group_1_1_Total_Exp`. I think you may have to use `Group_1_1_Separate_Exp` instead to get per-poly(A) site quantifications.

Saying that, I can also see that the `Group_1_1_Separate_Exp` column always has one fewer poly(A) sites than poly(A) sites reported in `Loci`. e.g. my `deapa` output for `control_r1` in the test data:

| Gene                             | Mean_Squared_Error | Predicted_APA       | Loci                  | Group_1_1_Separate_Exp | Group_1_1_Total_Exp |
|----------------------------------|--------------------|---------------------|-----------------------|------------------------|---------------------|
| ENSMUST00000023572.14\|NA\|16\|+ | 601.5              | 78340249,78340414   | 16:78336188-78340752  | 0.00,35.57,50.15       | 85.72               |
| ENSMUST00000088646.11\|NA\|6\|-  | 2449.3             | 119924719           | 6:119924622-119926687 | 14.76,147.08           | 161.84              |
| ENSMUST00000159903.1\|NA\|6\|-   | 2076.2             | 119924739,119924260 | 6:119923979-119926687 | 4.57,62.48,94.86       | 161.91              |
| ENSMUST00000177761.7\|NA\|6\|-   | 2076.2             | 119924739,119924260 | 6:119923979-119926687 | 4.57,62.48,94.86       | 161.91              |


I think that probably means the 3'end coordinate of the `Loci` column is the most distal poly(A) site and should also added to `apa` and reported in the output files.

Feel free to ask away if any of this is unclear :)

Reply:
@SamBryce-Smith Ah yes good catch! The separate expression column sounds like the right one to get TPM values from. I see, so if I understand correctly, we should get the following mapping from the first row above:

Position       Expression
78340249       0.00
78340414       35.57
78340752       50.15

Is this correct?

I can also contact the author about this later to double check our understanding just to be sure :)

Regarding this, since you mentioned that we should be using separate expression column instead, it just occurred to me that we'll get separate expression column for a different replicate of the same condition. Meaning if we have two replicates, we would get two expression values for each site. Should we then get the average of the two values to put in the final quantification challenge output file? I have also posted the same question on slack in case we need to discuss this further https://apaeval.slack.com/archives/C023SMTS0KS/p1631819510021800

Let me know your thoughts :D